qemu/tests/numa_prealloc_threads.py

```diff
@@ -43,27 +62,49 @@ def run(test, params, env):
     if not thread_id:
         test.fail("No thread-id setted.")
 
-    expected_cpu_affinity = thread_context_device.get_param("cpu-affinity")
+    error_context.context("Check the CPU affinity", test.log.info)
+    expected_cpu_affinity = thread_context_device.get_param("cpu-affinity", "0")
     cpu_affinity = vm.monitor.qom_get(thread_context_device_id, "cpu-affinity")
-    affinity = str(cpu_affinity[0]) + "-" + str(cpu_affinity[-1])
+    if cpu_affinity[0] != 0:
+        affinity = str(cpu_affinity[0]) + "-" + str(cpu_affinity[-1])
+    else:
+        affinity = "0"
     test.log.debug("The affinity: %s and the expected_cpu_affinity: %s"
                    % (affinity, expected_cpu_affinity))
     if expected_cpu_affinity != affinity:
         test.fail("Test and QEMU cpu-affinity does not match!")
 
-    cmd_taskset = "taskset -c -p " + str(thread_id)
-    output = process.getoutput(cmd_taskset)
-    if not re.search(affinity, output):
-        test.fail("The affinities %s and %s do not match!"
-                  % (affinity, str(output)))
+    cmd_taskset = "taskset -c -p %s" % thread_id
+    check_affinity(affinity, cmd_taskset, test)
 
     sandbox = params.get("qemu_sandbox", "on")
     if sandbox == "on":
-        try:
-            # The command is expected to fail
-            vm.monitor.qom_set(thread_context_device_id, "cpu-affinity", cpu_affinity)
-        except QMPCmdError as e:
-            test.log.debug("The expected error message: %s and the output: %s"
-                           % (error_msg, e.data))
-            if not re.search(error_msg, str(e.data)):
-                test.fail("Can not get expected error message: %s" % error_msg)
+        error_context.base_context("Setting a new affinity with sandbox enabled",
+                                   test.log.info)
+    try:
+        # The command is expected to fail when sandbox=on
+        vm.monitor.qom_set(thread_context_device_id,
+                           "cpu-affinity",
+                           first_cpu_affinity)
+        # Updates affinity with the latest value
+        affinity = str(first_cpu_affinity[0]) + "-" + str(first_cpu_affinity[-1])
+    except QMPCmdError as e:
+        error_context.context("The expected error message: %s and the output: %s"
+                              % (error_msg, e.data), test.log.debug)
+        if not re.search(error_msg, str(e.data)):
+            test.fail("Cannot get expected error message: %s" % error_msg)
```

Comment:
IMO, the logic for L81~L95 is incorrect.
1. L81~l83: `if sandbox=on:` do nothing but promote a log info.
2. The expect behaviour is set cpu-affinity success when sandbox off, but no such logic

Reply:
@yanan-fu for both sandbox cases the cpu-affinity has to be set.
For `sandbox=on` there's the log info, the cpu-affinity set and the exception handling
For `sandbox=off` only the affinity set (no log info and no exception as it's going to success)
What do you suggest to make these steps more clear?

Reply:
```
   sandbox = params.get_boolean("qemu_sandbox", "on")

    error_context.base_context("Setting cpu-affinity: %s" % first_cpu_affinity, test.log.info)
    try:
        vm.monitor.qom_set(thread_context_device_id, "cpu-affinity", convert_affinity(first_cpu_affinity))
    except QMPCmdError as e:
        if not sandbox:
            test.fail("Set cpu-affinity '%s' failed as: %s" % (first_cpu_affinity, str(e.data)))

        if not re.search(error_msg, str(e.data)):
            test.fail("Cannot get expected error message: %s" % error_msg)
        test.log.debug("Get the expected error message: '%s'" % error_msg)
    else:
        if sandbox:
            test.fail("Set cpu-affinity should fail when sandbox=on")
        cpu_affinity = first_cpu_affinity
```

Reply:
Updated, thanks a lot !